Refactoring the iterators

pwl opened the pull request:
~~I completely removed Options from Solver (there is no dictionary or options type).  For examples of how it works now see [test.jl](https://github.com/JuliaODE/ODE.jl/blob/dev/examples/test.jl).  The test's shouldn't pass at this stage.~~

This PR is a big refactor and simplification of the implementation.  New features include
- backward time integration for iterators
- `start`, `done` `next` replaced with `onestep!`, `init`, `trialstep!` and `errorcontrol!`
- the above functions return status (the particular form/type of status is still on a TODO list)

At the same time we removed
- rootfinding for dense stepper (temporarily)
- `points` keyword argument for dense output (it is an option only for `odeXX` interface)

pwl commented on src/types.jl:
```diff
@@ -116,123 +116,56 @@ function show(io::IO, state::Step)
-function collect(s::Solver)
-    if maximum(s.options.tspan) == Inf
-        error("Attempting to collect an infinite list, use tstop or tspan with finite numbers only")
+# Transpose a solver to get a (Vector(t),Vector(yout)) style output
```
This could also be implemented as `collect_transposed` removing the need for an additional type.

pwl committed:
Decoupling options and solver

pwl committed:
No more Iterators

pwl committed:
Removed the unused dtinit

mauro3 committed:
a few comments

mauro3 committed:
Dropped {T,Y} from Solver

pwl committed:
Merge pull request #5 from JuliaODE/m3/dev

a few comments

pwl committed:
Minor fixes after @mauro3 comments

pwl committed:
Adaptive and Fixed Options

pwl committed:
Tests for RK pass, removed backward integration

pwl committed:
Fixed ode23s

pwl committed:
Fixed the tests for the iterator interface

pwl committed:
Brought back the Jacobian

pwl committed:
New dense interface

pwl committed:
New backend and Runge-Kutta implementaion

pwl committed:
Updated docs

mauro3 committed:
some cleanup of iteration

mauro3 committed:
running partly again

mauro3 committed:
addressed comments

mauro3 committed:
Bug-dic in errorcontrol! in RK

mauro3 committed:
Added tdir

Iterator works, higher level functions don't yet.

mauro3 committed:
import Compat.String

mauro3 committed:
updated statuses

pwl committed:
Merge pull request #8 from JuliaODE/m3/dev

Added tdir

mauro3 committed:
ode23s mostly working

mauro3 committed:
Working ode23s

pwl commented on src/types.jl:
```diff
@@ -154,6 +155,8 @@ end
 Base.eltype{O}(::Type{Solver{O}}) = eltype(O)
 Base.eltype{O}(::Solver{O}) = eltype(O)
 
+tdir(s::Solver) = sign(s.stepper.options.tstop - s.ode.t0)
```
I don't think this should be a general function.  In general a stepper is not guaranteed to have options or tstop in particular.

pwl replied:
Maybe if/when we move `tstop` to solver we should generalize this.

pwl committed:
New dense output

mauro3 committed:
added example

mauro3 committed:
Merge branch 'dev' of github.com:JuliaODE/ODE.jl into jode/dev

pwl committed:
Brought interfaces.jl up do date with the new dense

pwl committed:
Reverse time integration for dense output

mauro3 commented on src/dense.jl:
```diff
@@ -1,193 +1,207 @@
+"""
+
+immutable DenseOptions{T<:Number} <: Options{T}
+    tspan    ::Vector{T}
```
`tspan` suggest an interval to me.  I think `tout` would be good.  (and `tsteps` for the fixed step solvers once I convinced you ...)

mauro3 commented on src/dense.jl:
```diff
@@ -1,193 +1,207 @@
+
+immutable DenseOptions{T<:Number} <: Options{T}
+    tspan    ::Vector{T}
+    tstop    ::T
```
`tstop` is in already in the Options, better remove here.

pwl commented on src/ode23s.jl:
```diff
@@ -78,76 +80,165 @@ function start{O<:ExplicitODE,S<:ModifiedRosenbrockStepper}(s::Solver{O,S})
+    # increment iteration counter
+    state.iters += 1
+    if state.iters > opts.maxiters
+        println("Reached maximum number of iterations $(opts.maxiters)")
```
maybe a `warn`?

pwl commented on src/ode23s.jl:
```diff
@@ -78,76 +80,165 @@ function start{O<:ExplicitODE,S<:ModifiedRosenbrockStepper}(s::Solver{O,S})
-        dtnew = min(opts.maxstep,
-                    dt*0.8*err^(-1/3) )
+     # allowable error
+    delta = max(opts.reltol*max(opts.norm(y), opts.norm(ynew)),opts.abstol)
```
This might turn out to be pretty slow without function types.

pwl replied:
But this is a fix for another day.

mauro3 replied:
I think we should optimize for 0.5, both the code and our time.  By the time this PR is merged 0.5 will be out.

pwl commented on src/ode23s.jl:
```diff
@@ -78,76 +80,165 @@ function start{O<:ExplicitODE,S<:ModifiedRosenbrockStepper}(s::Solver{O,S})
+    # update the state
+    step.t     = step.t+state.dtold
+    step.y[:]  = state.ynew
+    step.dy[:] = state.F2
```
`copy!`?

pwl commented on src/ode23s.jl:
```diff
@@ -78,76 +80,165 @@ function start{O<:ExplicitODE,S<:ModifiedRosenbrockStepper}(s::Solver{O,S})
+
+#     return tout, yout
+
+# end
```
Remove the commented code

mauro3 commented on src/dense.jl:
```diff
@@ -1,193 +1,207 @@
+immutable DenseOptions{T<:Number} <: Options{T}
+    tspan    ::Vector{T}
+    tstop    ::T
+    # points   ::Symbol
```
Aren't points needed?  Add TODO.

pwl replied:
I wrote about it on gitter

pwl replied:
But I can't find it so I will write here: if you wrap a solver with a dense output you aren't going to specify `points=:all` anyhow.  So why bother with this option here?  On the other hand, for `odeXX` you can specify `points=:specified` or `points=:all` to wrap a stepper in a dense output method or not.

mauro3 replied:
`:all` just means that it should return both the requested outputs and the times the steps were taken.  So a dense output should return those.

mauro3 committed:
Merge branch 'dev' of github.com:JuliaODE/ODE.jl into jode/dev

[ci skip]

mauro3 commented on src/dense.jl:
```diff
@@ -1,193 +1,207 @@
+output times stored in `options.tspan`).
+
+"""
+immutable DenseStepper{S<:Solver,O<:DenseOptions} <: AbstractStepper
```
Usually steppers are integrators, but this here is something else.  Confusing.  Does it need to be an `AbstractStepper` for any functionality?  Answer: no.

pwl committed:
Fixed some problems with dense output

mauro3 commented on src/dense.jl:
```diff
@@ -1,193 +1,207 @@
-    solver_state = start(solver)
-    ytmp = copy(y0)
-    return DenseState(step0, step1, t0-1, true, solver_state, ytmp, false)
+function init{O<:ExplicitODE,S<:DenseStepper}(s::Solver{O,S})
```
Needs a different name for same reason as above: `DenseStepper` is not an integrator, thus should not follow this protocol.

mauro3 commented on src/dense.jl:
```diff
@@ -1,193 +1,207 @@
-        if s.options.stopevent(t1,s1.y)
-            break
-        end
+function onestep!{O<:ExplicitODE,S<:DenseStepper}(s::Solver{O,S}, state::DenseState)
```
Again, different name.  I think using start, next, and done is probably ok.

mauro3 committed:
updated according to pwl's comments

[ci skip]

pwl committed:
Minor refactoring

pwl committed:
Merge branch 'dev' of github.com:JuliaODE/ODE.jl into dev

mauro3 committed:
updated REQUIRE

mauro3 committed:
Merge branch 'dev' of github.com:JuliaODE/ODE.jl into jode/dev

pwl committed:
Changed tspan to tout

pwl committed:
Merge branch 'dev' of github.com:JuliaODE/ODE.jl into dev

pwl committed:
Travis

pwl committed:
Updated the badges

pwl committed:
Updated the badges

mauro3 committed:
Tableau restructure

Moved RK-tableaus to RK-solvers

Marked files which define the core-interface in ODE.jl

mauro3 committed:
starting on rename

mauro3 commented on src/dense.jl:
```diff
@@ -1,193 +1,209 @@
+
+function solve{T,S<:DenseStepper}(ode::ExplicitODE{T},
+                                  ::Type{S};
+                                  method = RKStepperAdaptive{:rk45},
```
The `method = RKStepperAdaptive{:rk45}` is redundant as it is contained int the wrapped `solver`, or can be.  I think this interface needs some revision.

pwl replied:
`solver` in `DenseStepper` is constructed using `method`, so `DenseStepper` is not wrapping something that existed before.

mauro3 replied:
Hmm, maybe it would be clearer in general if the function `solve` was generically defined by calling the constructor of the `Stepper` and the stepper just defines an ordinary constructor?  I added this to the TODO #3.

pwl replied:
But this is another point I was thinking of making.  Maybe we should get rid of the convention with the custom constructors and move their contents directly to `solve`, which is already defined for each stepper.  Currently the declaration of the constructor looks awkward and complicated and `solve` is not doing anything at all apart from putting `stepper` and `ode` together.

mauro3 replied:
But normal constructors are the normal thing.  Whereas requiring to use `solve` as a constructor needs extra docs, etc.  So the move would be to have a very slim `solve`, only defined once. And have normal constructors for the steppers.

mauro3 replied:
Ok, I guess this is kinda the case already.  I'll let this sink in for a bit, sorry for the noise.

pwl replied:
But you have to pass optional arguments in some way, and with a normal constructor for stepper this is not possible.

mauro3 commented on src/dense.jl:
```diff
@@ -1,193 +1,209 @@
-        end
+function onestep!(ode::ExplicitODE,
+                  stepper::DenseStepper,
+                  state::DenseState)
```
should be `(ode::IVP,` no?

Use `dstate` for dense state, otherwise it get confusing.

pwl replied:
Right, we are not restricted to `ExplicitODE` now!  `dstate` seems good.  Could you update this in your #9?

mauro3 commented on src/dense.jl:
```diff
@@ -1,193 +1,209 @@
-    elseif t == step1.t
-        copy!(y,y1)
+"""
+function interpolate!{T,Y}(state::AbstractState,step_prev::Step{T,Y},tout::T,step_out::Step{T,Y})
```
I think the signature should be `interpolate!{T,Y}(step_out::Step{T,Y},dstate::DenseState,tout::T)`:
- if possible the in-place variable should be first (although, I'm not really bothered by this)
- I think it should take the `dstate`.  Some interpolants may want to store more extra stuff in `DenseState` than just the previous `Step`

pwl replied:
But then how are you going to overload this method for other solvers?

mauro3 committed:
ode->ivp

mauro3 committed:
Rest of the renames

mauro3 committed:
improved tableau situation in RK

mauro3 committed:
Fixed inconsistency in RK-tableau conversion

mauro3 committed:
Merge pull request #9 from JuliaODE/m3/rename

M3/rename

mauro3 committed:
Updated dense output to wrap the Integrator

mauro3 committed:
Added tdir to dense

mauro3 committed:
Big file rename.

Fixes #10
